### PR TITLE
Add coverage via `codecov`

### LINKED ISSUE #2702: Verify `coverage` is producing useful information
During #2690 I realized that the `coverage` call in Travis seems to run but doesn't produce any artifacts we can inspect. I want to verify what it's producing and maybe output them somewhere.
 
(this is mostly a note to myself as a new dev here; I'm sure this is working as intended, I just don't know what intent that is or if we're paying attention to it still)

### PR DESCRIPTION
**Pull request by joshuacwnewton**
#### GitHub

- [ ] I've given this PR a concise, self-descriptive, and meaningful title
- [ ] I've linked relevant issues in the PR body
- [ ] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [ ] I've applied a [release milestone](https://github.com/spinalcordtoolbox/spinalcordtoolbox/milestones) (major, minor, patch) in line with [Semantic Versioning guidelines](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Misc%3A-Creating-a-new-release#convention-for-naming-releases) 
- [ ] I've assigned a reviewer

#### PR contents

- [ ] I've consulted [SCT's internal developer documentation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki) to ensure my contribution is in line with any relevant design decisions
- [ ] I've added [relevant tests](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Programming%3A-Tests) for my contribution
- [ ] I've updated the [relevant documentation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Programming%3A-Documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description

This PR adds coverage reporting via `codecov`. Some quick links:

* [Sample Coverage Files view for PR commit](https://codecov.io/gh/spinalcordtoolbox/spinalcordtoolbox/tree/21bdc147bc66b6ea8ec5d64fe90987f7b884d7e3)
* [Codecov documentation](https://docs.codecov.com/docs)

## Linked issues

Fixes #2702.

### CODE REVIEW

**Comment by joshuacwnewton** [comment]
https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3463#issuecomment-880128949 is probably a little too verbose, I'd say? 

There are a lot of settings to tweak, though: https://docs.codecov.com/docs/pull-request-comments

File: .ci.sh [hunk 1]
```diff
@@ -39,16 +39,13 @@ run_tests() {
...
-  # NB: Coverage does not currently work: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/2702
+  # NB: Testing using example from https://github.com/codecov/example-python
   activate_venv_sct
-  pip install coverage
-  echo -ne "import coverage\ncov = coverage.process_startup()\n" > sitecustomize.py
-  echo -ne "[run]\nconcurrency = multiprocessing\nparallel = True\n" > .coveragerc
-  COVERAGE_PROCESS_START="$PWD/.coveragerc" COVERAGE_FILE="$PWD/.coverage" pytest
-  coverage combine
+  pip install pytest-cov
+  pytest --cov=spinalcordtoolbox --cov-config setup.cfg --cov-branch --cov-report=xml
```
Comment (joshuacwnewton) [inline .ci.sh:45]: One other thing I'd like to try is running the API tests and the CLI tests separately, then having separate coverage reports for each.

That way, we can know if a LOC is tested via a unit test, an integration test, or both. That distinction is pretty important! (I _think_ codecov is smart enough to give a fancy report for this.)

Comment (joshuacwnewton) [reply .ci.sh]: Huzzah! I've used codecov's "flags" to give us toggles for coverage from API tests and/or CLI tests:

Screenshot from 2021-07-14 20-09-27

(Maybe I should name these tags `testing/api` and `testing/cli`, though.)

**Comment by kousu** [comment]
I kinda think coverage is useless, maybe even counterproductive. The basic problem is that covering *lines* is like sticking a single fork in a cake: it just requires a single data point. Real coverage would mean covering *every possible state*, which is extremely hard to do in a non-functional language. Sometimes coverage sort of gets you to test that different inputs, by at least forcing you to give inputs that go through each case in all your if-else trees, but it's extremely limited.

Real coverage, IMO, means testing:

- the base case(s)
- the error cases
- one or two 'typical' cases
- the extreme cases

(I think of this like math intervals: your cases to test might be: x < 0, x == 0, x == 1, 1 < x < 10, 10 <= x < 20, 20 <= x, math.isnan(x), math.isinf(x))

but that's not what codecov does, because it's not smart enough to define what your cases are by looking at your code, all it can tell is what lines exist.

Here's some thinkpieces that are on my side that I googled to prop up my argument and make it look more impressive:

* https://blog.ploeh.dk/2015/11/16/code-coverage-is-a-useless-target-measure/
* https://reese.dev/code-coverage-is-useless/
* https://pedrorijo.com/blog/intro-mutation/
* https://it-benchmarking-journal.com/2017/10/7-software-testing-metrics-useless/
* https://dimitris-livas.squarespace.com/blog/2018/1/23/why-code-coverage-is-not-a-reliable-metric

Thinking about it, I'll grant that in a dynamic language like python, basic coverage is useful to catch the kind of typoes that aren't syntax errors, where other languages would use a typechecker instead.

**Comment by joshuacwnewton** [comment]
I largely agree that line-by-line coverage % isn't a very useful metric in its own right. Coverage doesn't tell the whole story; all it does it let you know you've covered a line with _some_ test, but doesn't tell you anything about the quality of those tests themselves.

That said, I think that most of those blog posts are trying to address the sort of cargo-culty dev-management audience who chase metrics as the end-all-be-all. But, in doing so, I think the posts throw the baby out with the bathwater; by rehashing the same point that coverage % is a misleading metric to try to target, they miss other legitimate uses where tools like codecov can still provide informative feedback. Coverage just has to be treated as _a_ tool in the toolbelt, not the _only_ tool.

[One of the linked blog posts](https://blog.ploeh.dk/2015/11/16/code-coverage-is-a-useless-target-measure/) highlights this pretty well:

> ### Code coverage is still useful
> 
> While it's dangerous to use code coverage for target setting, collecting coverage metrics can still be useful.
> 
> Some people use it to find areas where coverage is weak. There may be good reasons that some parts of a code base are sparsely covered by tests, but doing a manual inspection once in a while is a good idea. Perhaps you find that all is good, but you may also discover that a quality effort is overdue.
> 
> In some projects, I've had some success watching the code coverage trend. When I review pull requests, I first review the changes by looking at them. If the pull request needs improvement, I work with the contributor to get the pull request to an acceptable quality. Once that is done, I've already made my decision on merging the code, and then I measure code coverage. It doesn't influence my decision to merge, but it tells me about the trend. On some projects, I've reported that trend back to the contributor while closing the pull request. I wouldn't report the exact number, but I'd remark that coverage went up, or down, or remained the same, by 'a little', 'much', etc. The point of that is to make team members aware that testing is important.
> 
> Sometimes, coverage goes down. There are many good reasons that could happen. Watching how coverage evolves over time doesn't mean that you have to pounce on developers every time it goes down, but it means that if something looks odd, it may be worth investigating.

For SCT in particular, with our many CLI tests that run without checking results, coverage is still able to give me a better understanding of which parts of the codebase aren't as critical. It gets the gears going in my head: _"Oh, this branching path isn't covered. Why is that? What do the existing tests look like? Oh, huh. Our tests look representative of typical usage. Is this part of the codebase even useful, then? Why was this structured like this? What was the point of this if it isn't ever reached? Can it be refactored? etc."_ This applies doubly so during pull requests where we have more flexibility to restructure something.

Coverage is sort of like the checklist box in our [PR template](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/master/.github/PULL_REQUEST_TEMPLATE.md): 

> - [ ] I've added relevant tests for my contribution

You could definitely argue that this checkbox is just as useless as coverage. "Relevant tests"? What does that even mean? 

Yet, it can still act as a rough trigger to get people to start asking questions about what it is they're writing. Still, I'd want to make sure that contributors try to understand that testing is nuanced and context-dependent; a covered line can still be poorly tested, and an uncovered line can be acceptable.

**Comment by joshuacwnewton** [comment]
One other thing. I like the list you've posted here:

> * the base case(s)
> * the error cases
> * one or two 'typical' cases
> * the extreme cases

When I contributed to `scikit-learn`, they were really good about explicitly defining the various cases in the code itself. For example, take [this snippet of code](https://github.com/scikit-learn/scikit-learn/blob/3ae7c7615343bbd36acece57825d8b0d70fd9da4/sklearn/manifold/_t_sne.py#L759-L861) I worked on. That's 100+ lines of _just input validation_ to this function. Combining coverage with this sort of rigor ended results in a really good sense of which cases have tests written for them and which don't. 

Now, SCT isn't nearly at this level of rigor, but my hope is that by having coverage, it will nudge us to start thinking a little more about this sort of thing.

**Comment by jcohenadad** [comment]
> Here's 100% coverage in cynical diagram form:

omg, i love it

**Comment by kousu** [comment]
> > . On some projects, I've reported that trend back to the contributor while closing the pull request. I wouldn't report the exact number, but I'd remark that coverage went up, or down, or remained the same, by 'a little', 'much', etc. The point of that is to make team members aware that testing is important.

This reminds me of how youtube censored subscriber counts and started rounding to significant digits instead. Because it was causing too much drama and people focusing on the wrong thing. I wish codecov was like that. Maybe if it just gave a colour.

I'm satisfied with using coverage as a tool that we understand the limitations of.

File: .ci.sh (hunk 1 above)
Comment (joshuacwnewton) [reply .ci.sh]: > (Maybe I should name these tags testing/api and testing/cli, though.)

Done in f6412bc.

**Comment by RignonNoel** [comment]
@joshuacwnewton @kousu For me code coverage is a tool, and as for all the tool, saying that it's not perfect and can be missuses is not a correct reason to not use it at all when we are professional.

I'm 100% okay with you about this religion thinking that numbers are the only way to think. And I'm not saying that we should have a 100% coverage on this project, it would make no sense. But coverage is a great tool, and not having it is just an handicap for us since we will need to take decision without proper tool to help us analyze and understand our current state.

---

For me code coverage is required for multiples reason:

### 1/ Validate Python syntax

Python is not a compiled language and so we can't be sure that our code is correctly type without controlling that every line is executed at minimum one time.

### 2/ Having a clear idea of which part of the system lack tests
 
Having 10.000+ unit tests doesn't mean you tested all the important part of your system. A code coverage allow you to know what are the code section that are not covered and so you can make sure there is no important part in this list or plan some test tasks.

Also, code coverage can inform you of the number of time every line has been executed. It's a good way to know if your actual state make sense (important piece are more tested that nice-to-have side effect feature)

### 3/ Tests are part of the code, you need to respect them as well

Every time you make a change on the code it's interesting to know if the code coverage increase, decrease or is pretty much stable. It's a great and really easy way to politely ask the contributor to add some test case to cover his new changes.

### Bonus: The retro-analysis of tests

Sometimes you have a lot of tests, and when I add some code and want to cover it correctly (because I don't like manual testing even during my development phase) I like to just check the coverage system and see which test-case already test this section of the code:

 - I can detect which test maybe need some refactor EVEN if they don't fail (comments no more relevant with the new refactor)
 - I can see if maybe I can just adapt an actual test or if I need to create a new one
 - I can see if the section is maybe already not that well covered and decide to put the effort to test it well since I have the section in mind and my tools already setup for it

**Comment by mguaypaq** [comment]
It sounds like we're all on the same page:

- good tests are important;
- code coverage can be useful as a tool for thinking;
- code coverage would be harmful if we used it rigidly.

Given that, I was pretty neutral about whether or not it shows up automatically on PRs. But then I was suddenly convinced by @RignonNoel's first point, "Validate Python syntax": I think that alone is a good reason to be able to see, easily, whether each line of code has actually been run once.

(On second thought this might be caught by a linter, which I assume we're running already? But on third thought maybe the linter wouldn't catch things like typos in variable names, etc. Either way, adding code coverage sounds like a mildly good idea.)

**Comment by joshuacwnewton** [comment]
> (On second thought this might be caught by a linter, which I assume we're running already? But on third thought maybe the linter wouldn't catch things like typos in variable names, etc. Either way, adding code coverage sounds like a mildly good idea.)

Yep! Your suspicion is correct, @mguaypaq. We have a linting [GitHub Action](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/master/.github/workflows/lint_code.yml), but right now the linter only runs on changed lines (because otherwise it would be far too noisy for new PRs). So, I definitely agree that that's a pro of the coverage tool.

----

Overall, I think this is worth trying out! And we can always return to this topic once we've had a chance to work with it for a few weeks/months, too. :)

Would either of you (@RignonNoel or @mguaypaq) be able to review the code changes in this PR, then? This is my first time trying to set up codecov, and while l think I've got it set up well, it was also half a year ago, so there may be ways to configure codecov that I've missed. (e.g. the [`.codecov.yml`](https://docs.codecov.com/docs/codecov-yaml) docs.)

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -210,9 +210,22 @@ jobs:
...
         run: |
           ./.ci.sh -c
-      - name: Run pytest test suite
-        run: |
-          ./.ci.sh -t
+      - name: Run pytest test suite (with code coverage)
+        run: |  # NB: We use `-v` instead of `-t` here to run this one with coverage
+          ./.ci.sh -v
+      - name: Upload API coverage to Codecov
+        uses: codecov/codecov-action@v1
```
Comment (mguaypaq) [inline .github/workflows/tests.yml:217]: It looks like version 1 of `codecov-action` will stop working at the end of the month: [deprecation of v1](https://github.com/marketplace/actions/codecov#%EF%B8%8F--deprecration-of-v1). Thankfully, it looks like version 2 supports all of the flags we use ([docs](https://github.com/marketplace/actions/codecov#arguments)):
- fail_ci_if_error
- files
- flags

**Review by mguaypaq — APPROVED** [review]
I haven't worked with codecov before, but from looking at the documentation, this looks like a fine configuration to start with. I would be comfortable if you merged this after addressing the comments below.

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -210,9 +210,22 @@ jobs:
       - name: Check dependencies
         run: |
           ./.ci.sh -c
-      - name: Run pytest test suite
-        run: |
-          ./.ci.sh -t
+      - name: Run pytest test suite (with code coverage)
+        run: |  # NB: We use `-v` instead of `-t` here to run this one with coverage
```
[the remark this answers is not in the document]
Comment (joshuacwnewton) [reply .github/workflows/tests.yml:214]: Ah, thank you for pointing this out! 

Just to provide some context, the `batch_processing` tests act as a much bigger integration test, chaining multiple `sct` commands together in a pipeline shell script. This requires about ~15-30m of dedicated processing before any results can actually be validated. So, to save time, we run them in a separate CI workflow, parallel to the more granular unit tests.

It's not very clear by just looking at the folders (sorry about that), but the tests in `testing/batch_processing` will only be run if a specific environment variable is set first. And, we set that environment variable via the [`test-batch-processing.yml`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/master/.github/workflows/test-batch-processing.yml) CI workflow that manages the `batch_processing` tests.

So, right now `-t` will run the same tests as `-v` (even though that isn't very clear!)

[the remark this answers is not in the document]
Comment (joshuacwnewton) [reply .github/workflows/tests.yml:214]: Typing this out, though, it isn't very clear that batch_processing gets skipped by `-v`. There might be a better way of  organizing things so that it is more apparent. :thinking:

[the remark this answers is not in the document]
Comment (joshuacwnewton) [reply .github/workflows/tests.yml:214]: Perhaps we could update `-t` to explicitly run `pytest` on `testing/api` and `testing/cli` via 2 separate commands, too? That way, if we _do_ add tests in a new folder, we'll be forced to make updates to account for the new folder, and thus will be less likely to silently miss running a test.

File: .ci.sh [hunk 4]
```diff
@@ -39,16 +39,13 @@ run_tests() {
...
-  # NB: Coverage does not currently work: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/2702
+  # NB: Testing using example from https://github.com/codecov/example-python
   activate_venv_sct
-  pip install coverage
-  echo -ne "import coverage\ncov = coverage.process_startup()\n" > sitecustomize.py
-  echo -ne "[run]\nconcurrency = multiprocessing\nparallel = True\n" > .coveragerc
-  COVERAGE_PROCESS_START="$PWD/.coveragerc" COVERAGE_FILE="$PWD/.coverage" pytest
-  coverage combine
+  pytest --cov=spinalcordtoolbox --cov-config setup.cfg --cov-branch --cov-report=xml:cov-api.xml testing/api
+  pytest --cov=spinalcordtoolbox --cov-config setup.cfg --cov-branch --cov-report=xml:cov-cli.xml testing/cli
```
[the remark this answers is not in the document]
Comment (joshuacwnewton) [reply .ci.sh:45]: Re: the context given in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3463#discussion_r786162517, it might be a bit tricky to get coverage, since `pytest` is only used to validate the end results of the pipeline. In other words, `pytest` isn't used to invoke the processing itself, so I'm not sure if we can hook into any line-by-line coverage scanning. :(

File: .github/workflows/tests.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (mguaypaq) [reply .github/workflows/tests.yml:214]: Thanks for the context. Changing `-t` to be explicit like `-v` would be clearer, yes. Although it doesn't have to be two separate commands, just `pytest testing/api testing/cli` would be fine.

File: .github/workflows/tests.yml (hunk 2 above)
Comment (joshuacwnewton) [reply .github/workflows/tests.yml:217]: Ah, good catch! I didn't even realize there was a distinction between v1 and v2. Thank you for noticing this. :D
